### PR TITLE
fixes for various networking/ATP stuck download failure cases

### PR DESCRIPTION
**Pull request by birarda**
- fixed an issue with stuck AssetClient requests made after adding Asset Server but before activating Asset Server socket
- fixed an issue with multiple Asset Servers being added to Interface NodeList (Asset Server should have been a solo node like Audio Mixer and Avatar Mixer)
- fixed an issue with lost/stuck asset requests on Asset Server quick disconnect/reconnect
- added faked handshake function to AssetScriptingInterface in PR_BUILD and DEV_BUILD
- fixed an issue with udt::Connections created for nodes no longer present in NodeList
- fixed an issue with udt::SendQueue timeout when no response had ever been received

### CODE REVIEW

**Comment by birarda** [comment]
This PR fixes 1885 and 1818, including a couple of additional issues found while debugging those.

Will have test notes tomorrow morning at latest. They will be complicated, this covers a variety of odd edge cases that need a very specific setup to repro.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 30259a30eed91cea2a66e3826a7b6c7a5aa0eb36 available at: https://gist.github.com/1403f877898eb1878e600a250d136e1f<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8797-30259a3.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8797-30259a3.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 9684357fa022f745d642239c173c248ecd53e4b8 available at: https://gist.github.com/1403f877898eb1878e600a250d136e1f<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8797-9684357.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8797-9684357.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: fb65260162e4d03638db2c79dc397e1303558480 available at: https://gist.github.com/1403f877898eb1878e600a250d136e1f<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8797-fb65260.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8797-fb65260.dmg

**Comment by birarda** [comment]
#### Test Plan

At least two of these issues are difficult to reproduce since they are races. You may need to try the repro case a number of times before you catch it.

Use [the PR 8804 build](https://github.com/highfidelity/hifi/pull/8804) as the "bad" Interface/Sandbox to reproduce the issues explained below. 

Use this PR build to test that the reproduced issues are eliminated.

##### Fix for rapid disconnect and reconnect while loading content

You will want expanded stats visible (display with / and click on the displayed stats to expand) to track Asset download bandwidth and pending stuck downloads.
 
1. Use Interface from PR 8804 connected to your Sandbox (localhost).
2. Click menu option to reload all content. 
3. Immediately once content has begun reload quickly jump to another domain and then go back to localhost. The fastest way to do this is via the discovery/address bar using the discovery thumbnails to jump away and the home icon to jump back.
4. Repeat steps 2-3, reloading content to ensure there are ongoing downloads and hopping off and localhost and back.

Eventually Interface should be stuck in a state of very slow downloading (the display Asset bandwidth in will drop and downloads will take a long time to complete).

Repeat the steps above a number of times with this PR build and confirm you do not have the issue.

##### Fix for multiple Asset Servers in NodeList

1. Use Interface from PR 8804 connected to your Sandbox (localhost).
2. Figure out the process ID for your Asset Server. The easiest way is to check the Sandbox logs and look at the PID in the log for the Asset Server.
3. Using task manager, end the assignment-client.exe process for your Asset Server.
4. Observe in Interface that your number of servers shown in stats climbs to 6 before going back down to 5.

Repeat the steps above with this PR build. Your number of servers shown in stats should never climb to 6.

##### Fix for handshake race

1. Use Interface from PR 8804 connected to your Sandbox (localhost).
2. Prepare a script console window with the following command `Assets.sendFakedHandshake()` but do not send the command.
3. Reload all content.
4. While content is reloading, use the script console to call `Assets.sendFakedHandshake()` multiple times. 
5. You may need to repeat steps 2-4 multiple times.

Eventually you should have a couple of downloads in the list in stats that are completely stuck that will never complete.

Perform the same test with this PR build, it should be impossible to end up with stuck downloads.

**Comment by OmegaHeron** [comment]
Working on this - expect it to take a few hours

File: libraries/networking/src/LimitedNodeList.cpp [hunk 1]
```diff
@@ -1136,3 +1144,21 @@ void LimitedNodeList::flagTimeForConnectionStep(ConnectionStep connectionStep, q
...
+void LimitedNodeList::clientConnectionToSockAddrReset(const HifiSockAddr& sockAddr) {
+    // for certain reliable channels higher level classes may need to know if the udt::Connection has been reset
+    auto matchingNode = findNodeWithAddr(sockAddr);
+
+    if (matchingNode) {
+        emit clientConnectionToNodeReset(matchingNode);
+    }
+}
+
+void LimitedNodeList::sendFakedHandshakeRequestToNode(SharedNodePointer node) {
```
Comment (ZappoMan) [inline libraries/networking/src/LimitedNodeList.cpp]: should this be in `#if (PR_BUILD || DEV_BUILD)` to prevent misuse?

**Comment by OmegaHeron** [comment]
I believe this is behaving as expected - despite my best efforts I could not break it.   This + Vegas should resolve a lot of evils.

File: libraries/networking/src/LimitedNodeList.cpp (hunk 1 above)
Comment (birarda) [inline libraries/networking/src/LimitedNodeList.cpp]: It isn't exposed anywhere usable except in PR_BUILD and DEV_BUILD so I misuse isn't really possible, but it might as well be guarded the same way if that's the only way we use it.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 3e8b14bc627964d8cec5574c6b55a626edc56e8e available at: https://gist.github.com/1403f877898eb1878e600a250d136e1f<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8797-3e8b14b.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8797-3e8b14b.dmg

**Comment by birarda** [comment]
Marking this DNM, has a bug with ATP scripts for agent assignment-clients.

**Comment by birarda** [comment]
#### Test Plan for 141c164

A special extra test plan for the fix introduced in `141c164`. Test the repro/fix against build for `3e8b14b`.

On `3e8b14b`, AC scripts for agent ACs will fail to download over ATP. The AC will never get the script to run and will keep restarting trying to get the script and failing.

This is fixed in `141c164`. Verify that `141c164` still does not have the problems explained in *Fix for handshake race* in the first test plan (namely that you are unable to get stuck Asset downloads if you send a faked handshake).

**Comment by howard-stearns** [comment]
works well in my tests. Ran domain-check.js 3x against new-home/dev-welcome/dev-playa, then restarted and ran 3x more. No problems observed.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 141c164ebc111dca6f0a2a238e1bc15700ff1756 available at: https://gist.github.com/1403f877898eb1878e600a250d136e1f<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8797-141c164.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8797-141c164.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 3a1eda25ea8e201a239646d49bccb2415cd0ec53 available at: https://gist.github.com/1403f877898eb1878e600a250d136e1f<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8797-3a1eda2.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8797-3a1eda2.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 9184a500a093dfda23fd739e43a3a5a051060564 available at: https://gist.github.com/1403f877898eb1878e600a250d136e1f<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8797-9184a50.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8797-9184a50.dmg

File: libraries/networking/src/AssetClient.cpp [hunk 2]
```diff
@@ -45,6 +45,8 @@ AssetClient::AssetClient() {
     packetReceiver.registerListener(PacketType::AssetUploadReply, this, "handleAssetUploadReply");
 
     connect(nodeList.data(), &LimitedNodeList::nodeKilled, this, &AssetClient::handleNodeKilled);
+    connect(nodeList.data(), &LimitedNodeList::clientConnectionToNodeReset,
+            this, &::AssetClient::handleNodeClientConnectionReset);
```
Comment (Atlante45) [inline libraries/networking/src/AssetClient.cpp]: Why `&::AssetClient`?

Comment (birarda) [inline libraries/networking/src/AssetClient.cpp]: Whoops

**Comment by hifi-gustavo** [comment]
Build console output for commit: d1cd99e035d788aaafd94a33403390481a8ac137 available at: https://gist.github.com/1403f877898eb1878e600a250d136e1f<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8797-d1cd99e.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8797-d1cd99e.dmg

**Comment by RyanDowne** [comment]
I think this is ok to mark QA approved?

**Comment by Atlante45** [comment]
@RyanDowne Oh yeah it is, I had not seen Omega's comment. Merging.
